### ISSUE TITLE
Rule proposal: No promise inside logical expression

### ISSUE THREAD

**Issue by the21st** [typescript-eslint/typescript-eslint#3403]
### Please describe what the rule should do:

The rule should warn about putting a `Promise` inside a logical expression because it always resolve to true.

In most of the cases, we want to use the result of the promise instead.

### What category of rule is this? (place an "X" next to just one item)

[X] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Enforces code style (layout)
[ ] Other (please specify:)

Provide 2-3 code examples that this rule will warn about:

```ts
async function isValid() {
  return false;
}

const firstCheck = true
const allChecks = firstCheck && isValid() // This is most likely an error because promise always resolves to true

if (allChecks) {
  // Do work
}
```

Correct code:
```ts
const allChecks = firstCheck && (await isValid())
```

Why should this rule be included in ESLint (instead of a plugin)?

This is a common mistake instead of code style. I think it should be included in ESLint to avoid this mistake.

**Comment by bradzacher** [comment]
Have you tried the `no-misused-promise` lint rule? It should do what you want.

Additionally new versions of typescript will do this out of the box if your variable is types as `Promise<foo>`

**Comment by the21st** [comment]
Hi @bradzacher, thanks for the quick reply!

I have the `no-misused-promises` rule turned on (`'@typescript-eslint/no-misused-promises': 'error'`) but it's not triggering anything on the code sample I posted.

I'm on Typescript 4.2.3 right now, can you please provide more context about its check for this? I have strict mode on. Or is it an upcoming feature in TS 4.3?

Thanks!

**Comment by bradzacher** [comment]
> I have the no-misused-promises rule turned on ('@typescript-eslint/no-misused-promises': 'error') but it's not triggering anything on the code sample I posted.

Then likely your code sample does not actually mirror the code you have in your project.

As you can see in this repro - the rule correctly reports.
https://yeonjuan.github.io/typescript-eslint-demo/#e2NvZGU6J2FzeW5jIGZ1bmN0aW9uIGlzVmFsaWQoKSB7XG4gIHJldHVybiBmYWxzZTtcbn1cblxuY29uc3QgZmlyc3RDaGVjayA9IHRydWVcbmNvbnN0IGFsbENoZWNrcyA9IGZpcnN0Q2hlY2sgJiYgaXNWYWxpZCgpIC8vIFRoaXMgaXMgbW9zdCBsaWtlbHkgYW4gZXJyb3IgYmVjYXVzZSBwcm9taXNlIGFsd2F5cyByZXNvbHZlcyB0byB0cnVlXG5cbmlmIChhbGxDaGVja3MpIHtcbiAgLy8gRG8gd29ya1xufVxuJyxydWxlczp7J0B0eXBlc2NyaXB0LWVzbGludC9uby1taXN1c2VkLXByb21pc2VzJzonZXJyb3InfSxwYXJzZXJPcHRpb25zOntlY21hRmVhdHVyZXM6e2pzeDpmYWxzZSxnbG9iYWxSZXR1cm46ZmFsc2V9LGVjbWFWZXJzaW9uOjIwMTgscHJvamVjdDpbJy4vdHNjb25maWcuanNvbiddLHNvdXJjZVR5cGU6J3NjcmlwdCd9fQ==

> is it an upcoming feature in TS 4.3?

yes it is

https://www.typescriptlang.org/play?ts=4.3.0-beta&noFallthroughCasesInSwitch=true&noPropertyAccessFromIndexSignature=false&strict=true#code/IYZwngdgxgBAZgV2gFwJYHsI1SAasAG1QBMAKAShgG8BYAKBhgCcBTZBJrOQkFgbnoBfevSiYQyeKiYSAwgAsWUANYwAvDGRMELUeMmECCpcpDqpM5MZUwAZLex5CJCjAD0bmABV5ORzABbdAkYImUWAjAYYCwWJiZ0JhgAIyVgBF4YAAcEgJwWaIIAd2AwM1YQdAIANxYzZHRNbV06elQ4GFJDa1NKWgZ3TwARRqLE5SF6IA

**Comment by the21st** [comment]
@bradzacher you're right, thanks for double-checking my example. I stripped down my actual code to what I thought was the minimum example but something was lost in that process I didn't expect.

This is the actual code:

```ts
async function isValid() {
  return false;
}

interface User {
  isActive: boolean
}

function getUser(): User | undefined {
  return undefined
}

const user = getUser()

const allChecks = user?.isActive && isValid()

if (allChecks) {
  // Do work
}
```

Neither TS4.3 nor `no-misused-promises` triggers anything. Do you think it's a valid example for `no-misused-promises` ?

**Comment by bradzacher** [comment]
the problem here is that it's impossible to tell your meaning.

`const allChecks = user?.isActive && isValid()` is perfectly valid code - there's nothing wrong with assigning a promise to a variable. It's actually a very common thing to do. For example

```ts
const promiseOne = doSomethingAsync();
const promiseTwo = doSomethingAsync();

// do some other work

await Promise.all([promiseOne, promiseTwo]);
```

The logical expression doesn't make it any less valid as it's ofc short-circuiting the triggering of the promise.

---

`allChecks` is typed as `false | Promise<boolean> | undefined` - so again there's nothing analysably wrong with your code.

From a static analysis POV, the question being asked is this:
"was the `if` intended to narrow the type to the truthy `Promise<boolean>`, or was the intention to resolve the boolean as part of the check?"

The answer to that question is "it could be either - only the original engineer knows for sure".

For a human, ofc you can look at the code, follow the assignments of `allChecks` and determine if it's ever `await`ed, and then decide what the intention is - but unfortunately it's *hard* for static analysis to do that.

---

The problem here really is the variable assignment. It makes the code ambiguous to static analysis.

If you inline the variable within the `if`, then `no-misused-promises` can statically understand your meaning and tell that your promise is 100% used incorrectly (though it looks like TS4.3 doesn't error on this)

```ts
if (user?.isActive && isValid()) {
  // Do work
}
```
[repl](https://yeonjuan.github.io/typescript-eslint-demo/#e2NvZGU6J2FzeW5jIGZ1bmN0aW9uIGlzVmFsaWQoKSB7XG4gIHJldHVybiBmYWxzZTtcbn1cblxuaW50ZXJmYWNlIFVzZXIge1xuICBpc0FjdGl2ZTogYm9vbGVhblxufVxuXG5mdW5jdGlvbiBnZXRVc2VyKCk6IFVzZXIgfCB1bmRlZmluZWQge1xuICByZXR1cm4gdW5kZWZpbmVkXG59XG5cbmNvbnN0IHVzZXIgPSBnZXRVc2VyKClcblxuaWYgKHVzZXI/LmlzQWN0aXZlICYmIGlzVmFsaWQoKSkge1xuICAvLyBEbyB3b3JrXG59XG4nLHJ1bGVzOnsnQHR5cGVzY3JpcHQtZXNsaW50L25vLW1pc3VzZWQtcHJvbWlzZXMnOidlcnJvcid9LHBhcnNlck9wdGlvbnM6e2VjbWFGZWF0dXJlczp7anN4OmZhbHNlLGdsb2JhbFJldHVybjpmYWxzZX0sZWNtYVZlcnNpb246MjAxOCxwcm9qZWN0OlsnLi90c2NvbmZpZy5qc29uJ10sc291cmNlVHlwZTonc2NyaXB0J319)

**Comment by the21st** [comment]
@bradzacher wow, thanks for the extensive explanation! I also appreciate you created an issue in the TypeScript repo. Feel free to close this issue now that there's a new one there.

>The problem here really is the variable assignment.

I'm not sure why variable assignment is a problem though. Taking the inside of an `if` statement and assigning it to a named variable doesn't really change the semantics of the code, I always considered this more of documentation step (giving the condition a name that makes sense).

**Comment by the21st** [comment]
Thanks for the insights! I don't really know how eslint works under the hood, so I appreciate you taking the time to explain. 

Coming from a background of mostly statically-typed languages, I assumed typescript-eslint mostly looked at types. In such case, `var` and `condition && promise` would be the same, because both evaluated to the same type. It's actually a bit surprising to me this is not how it works. Thank you! 🙂
